File: frontends/client-portal/src/types/supportTickets.ts
export type SupportTicketStatus = "OPEN" | "IN_PROGRESS" | "CLOSED";

export type SupportTicketPriority = "LOW" | "NORMAL" | "HIGH";

export type SupportTicketSlaStatus = "OK" | "BREACHED" | "PENDING";

export interface SupportTicketItem {
  id: string;
  org_id: string;
  created_by_user_id: string;
  subject: string;
  message: string;
  status: SupportTicketStatus;
  priority: SupportTicketPriority;
  first_response_due_at: string | null;
  first_response_at: string | null;
  resolution_due_at: string | null;
  resolved_at: string | null;
  sla_first_response_status: SupportTicketSlaStatus;
  sla_resolution_status: SupportTicketSlaStatus;
  sla_first_response_remaining_minutes: number | null;
  sla_resolution_remaining_minutes: number | null;
  case_id?: string | null;
  case_status?: "TRIAGE" | "IN_PROGRESS" | "WAITING" | "RESOLVED" | "CLOSED" | null;
  case_queue?: "FRAUD_OPS" | "FINANCE_OPS" | "SUPPORT" | "GENERAL" | null;
  case_priority?: "LOW" | "MEDIUM" | "HIGH" | "CRITICAL" | null;
  case_updated_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface SupportTicketComment {
  user_id: string;
  message: string;
  created_at: string;
}

export interface SupportTicketDetail extends SupportTicketItem {
  comments: SupportTicketComment[];
}

export interface SupportTicketListResponse {
  items: SupportTicketItem[];
  next_cursor?: string | null;
}

export interface SupportTicketCreatePayload {
  subject: string;
  message: string;
  priority: SupportTicketPriority;
}

export interface SupportTicketCommentPayload {
  message: string;
}

File: platform/processing-core/app/schemas/support_tickets.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from app.models.support_ticket import SupportTicketPriority, SupportTicketSlaStatus, SupportTicketStatus


class SupportTicketCreate(BaseModel):
    subject: str = Field(..., min_length=1, max_length=255)
    message: str = Field(..., min_length=1)
    priority: SupportTicketPriority = SupportTicketPriority.NORMAL


class SupportTicketCommentCreate(BaseModel):
    message: str = Field(..., min_length=1)


class SupportTicketCommentOut(BaseModel):
    user_id: str
    message: str
    created_at: datetime


class SupportTicketAttachmentInit(BaseModel):
    file_name: str = Field(..., min_length=1, max_length=255)
    content_type: str = Field(..., min_length=1, max_length=128)
    size: int = Field(..., ge=1)


class SupportTicketAttachmentComplete(BaseModel):
    object_key: str = Field(..., min_length=1)
    file_name: str = Field(..., min_length=1, max_length=255)
    content_type: str = Field(..., min_length=1, max_length=128)
    size: int = Field(..., ge=1)


class SupportTicketAttachmentInitResponse(BaseModel):
    upload_url: str
    object_key: str


class SupportTicketAttachmentOut(BaseModel):
    id: str
    ticket_id: str
    org_id: str
    uploaded_by_user_id: str
    file_name: str
    content_type: str
    size: int
    object_key: str
    created_at: datetime


class SupportTicketAttachmentListResponse(BaseModel):
    items: list[SupportTicketAttachmentOut]


class SupportTicketOut(BaseModel):
    id: str
    org_id: str
    created_by_user_id: str
    subject: str
    message: str
    status: SupportTicketStatus
    priority: SupportTicketPriority
    first_response_due_at: datetime | None
    first_response_at: datetime | None
    resolution_due_at: datetime | None
    resolved_at: datetime | None
    sla_first_response_status: SupportTicketSlaStatus
    sla_resolution_status: SupportTicketSlaStatus
    sla_first_response_remaining_minutes: int | None
    sla_resolution_remaining_minutes: int | None
    created_at: datetime
    updated_at: datetime


class SupportTicketDetail(SupportTicketOut):
    comments: list[SupportTicketCommentOut] = Field(default_factory=list)


class SupportTicketListResponse(BaseModel):
    items: list[SupportTicketOut]
    next_cursor: str | None = None


__all__ = [
    "SupportTicketCommentCreate",
    "SupportTicketCommentOut",
    "SupportTicketAttachmentComplete",
    "SupportTicketAttachmentInit",
    "SupportTicketAttachmentInitResponse",
    "SupportTicketAttachmentListResponse",
    "SupportTicketAttachmentOut",
    "SupportTicketCreate",
    "SupportTicketDetail",
    "SupportTicketListResponse",
    "SupportTicketOut",
]
